nixos/modules/services/networking/cgit.nix

```diff
@@ -165,29 +175,66 @@ in
       message = "Exactly one of services.cgit.${vhost}.scanPath or services.cgit.${vhost}.repos must be set.";
     }) cfgs;
 
-    services.fcgiwrap.enable = true;
+    warnings = flatten (flip mapAttrsToList cfgs (inst: cfg:
+      optional (cfg.user == "root") ''
+        `services.cgit.${inst}` is configured to run as root.
+        This has security implications: <TODO: advisory link>.
```

Comment:
Thanks for your work on this. I think linking to the advisory is great; we should certainly post it on Discourse and arguably https://github.com/NixOS/nixpkgs/security/advisories (as‐yet unused).

I personally feel that a warning is not sufficient and that insecure configurations should be a hard error, because otherwise it’s likely they won’t be surfaced through automatic update pipelines etc.; however I realize there is a compatibility cost here.

Reply:
> I personally feel that a warning is not sufficient and that insecure configurations should be a hard error, because otherwise it’s likely they won’t be surfaced through automatic update pipelines etc.; however I realize there is a compatibility cost here.

I think the biggest issue with that, especially on the stable channel, is that for people using the auto-upgrade and not really monitoring their machine, it would block other compatible security upgrades. Like for example the recent curl CVE.

Reply:
There’s lots of circumstances where that would bite you. Packages *do* break unexpectedly. You need monitoring of failures to have security; Nix warnings, on the other hand, are rarely security‐relevant, and are trickier to automatically detect and monitor.

Reply:
> There’s lots of circumstances where that would bite you. Packages _do_ break unexpectedly. You need monitoring of failures to have security; Nix warnings, on the other hand, are rarely security‐relevant, and are trickier to automatically detect and monitor.

I agree, you definitely should monitor your machines. But I think lots of people still don't.

But it's true that warnings would be harder to detect, hmm…

Reply:
Perhaps we could emit a warning *and make the systemd services fail to start at runtime* with the same error. That way we fail closed in terms of this particular security vulnerability, anyone with any kind of monitoring (even manual) of “does the stuff I use even work at all” will notice that something’s amiss, and other security updates won’t be blocked.

Reply:
Hmmm, interesting. But note that this might be more severe than an eval failure, since in the case of an eval failure, the previously built system still continues running. What about making another dummy systemd service that fails? That way at least they monitor failing services, they'll be notified, without breaking anything.

Reply:
There really should be a proper notification channel for these kinds of things…

Reply:
Well, the system would continue running, just not the part of it with a privilege escalation vulnerability…

Fundamentally we have to do *something* that people will notice to get their attention about a security vulnerability if we don’t want to leave many hosts vulnerable to it.

Reply:
Yeah, which is why I think having a proper security ML or something would be good for these exact cases.

Right now we have to choose whether removing the vulnerability is more important than keeping the service running. In this instance, one of the affected services is zoneminder, which can be used to make automatic video-surveillance. Is a local privilege escalation more important than the user's video-surveillance?

One thing I'm sure is that I don't like being in that position :sweat_smile:

Reply:
Well, I’d still be just fine with failing the build if I’ve talked you into it!

Seriously though, a warning is better than nothing. But I do think we can do better.

By the way, you can subscribe to email updates from the [Announcements → Security](https://discourse.nixos.org/c/announcements/security/56) category on the Discourse; that’s effectively the advisories mailing list. Discoverability is another thing, though, as is making sure everything important actually gets posted there.

Reply:
> Well, I’d still be just fine with failing the build if I’ve talked you into it!

👎👎👎

If you feel that strongly about it, why not invest that energy into giving better visibility to warnings? E.g. by having autoUpdate send emails?

There are literally thousands of local privilege escalations on any given linux. Why not hunt for those instead of harrassing stable users?

Reply:
A warnings/notification system that actually works is an important thing that we’re missing across the entirety of Nixpkgs and NixOS. It would be a big improvement, and there has already been some work on similar things in terms of the problems infrastructure for packages. However, it would take a lot more energy and time than giving advice on this issue.

I think “literally thousands” is an exaggeration. Linux platform security is not in a great state, but it is not *that* bad, especially in a server environment. The existence of other vulnerabilities shouldn’t stop us correctly handling this one.

I reject the idea that letting someone know their system lets *any local user* escalate to root constitutes “harassment”. I think you have made your position very clear by now, however.

Reply:
Throwing an evaluation error is pretty much exactly the same behaviour as marking packages with `knownVulnerabilities`, which is a normal thing we do all the time to let people know that they are depending on vulnerable software, including in the stable branch. If we want to decide we should do something more nuanced, then it should address the far more common case of `knownVulnerabilities` first. In the meantime, we already have established precedent that we block builds of known‐insecure configurations by default and that you need to monitor failed updates to ensure you get updates (as of course you would have to anyway). This does not impact availability because failing to upgrade the system does not take down the existing one.

Reply:
I think we could just add a `services.fcgiwrap.allowLocalPrivilegeEscalationUntilNextStableRelease` option or something if we want an escape hatch.

I agree that the systemd service failure option is not ideal.

The GitHub advisory tab lets us assign CVEs, as I understand it, so it would actually be quite nice to use. But only repository admins can post stuff; we’d probably need to organize procedures and someone to be responsible for it.

Reply:
GHA are not really adequate in our situation with the current way the permissions are handled. Creation and publication needs to be handled by repo admins and cannot be delegated :/ (also not sure if tying more things with the GH ecosystem is a good play here).

In this specific case at this point the cat is out of the bag so I suggest to go ahead and publish the advisory on https://discourse.nixos.org/c/announcements/security/56 when this is ready to be merged to get the final link so it can be added to the messages.

Reply:
I think Discourse will be okay, but there’s indeed a bit of a race condition there. Maybe we can post something set to private and then let it go public once we merge? I’m not sure who has that kind of access on Discourse other than the moderation team.